python/lsst/daf/butler/datastores/posixDatastore.py

```diff
@@ -292,67 +293,139 @@ def put(self, inMemoryDataset, ref):
+        possibly transferring it.
 
-    def ingest(self, path, ref, formatter=None):
-        """Record that a Dataset with the given `DatasetRef` exists in the store.
+        The caller is responsible for ensuring that the given (or predicted)
+        Formatter is consistent with how the file was written; `ingest` will
+        in general silently ignore incorrect formatters (as it cannot
+        efficiently verify their correctness), deferring errors until ``get``
+        is first called on the ingested dataset.
 
         Parameters
         ----------
         path : `str`
-            File path, relative to the repository root.
+            File path.  Treated as relative to the repository root if not
+            absolute.
         ref : `DatasetRef`
             Reference to the associated Dataset.
         formatter : `Formatter` (optional)
-            Formatter that should be used to retreive the Dataset.
+            Formatter that should be used to retreive the Dataset.  If not
+            provided, the formatter will be constructed according to
+            Datastore configuration.
+        transfer : str (optional)
+            If not None, must be one of 'move', 'copy', 'hardlink', or
+            'symlink' indicating how to transfer the file.  The new
+            filename and location will be determined via template substitution,
+            as with ``put``.  If the file is outside the datastore root, it
+            must be transferred somehow.
+
+        Raises
+        ------
+        RuntimeError
+            Raised if ``transfer is None`` and path is outside the repository
+            root.
+        FileNotFoundError
+            Raised if the file at ``path`` does not exist.
+        FileExistsError
+            Raised if ``transfer is not None`` but a file already exists at the
+            location computed from the template.
         """
         if formatter is None:
             formatter = self.formatterFactory.getFormatter(ref.datasetType.storageClass,
                                                            ref.datasetType.name)
-        # Create Storage information in the registry
-        ospath = os.path.join(self.root, path)
-        checksum = self.computeChecksum(ospath)
-        stat = os.stat(ospath)
-        size = stat.st_size
-        info = StorageInfo(self.name, checksum, size)
-        self.registry.addStorageInfo(ref, info)
 
-        # Associate this dataset with the formatter for later read.
-        fileInfo = StoredFileInfo(formatter, path, ref.datasetType.storageClass)
-        self.addStoredFileInfo(ref, fileInfo)
-
-        # Register all components with same information
-        for compRef in ref.components.values():
-            self.registry.addStorageInfo(compRef, info)
-            self.addStoredFileInfo(compRef, fileInfo)
+        with self.transaction() as transaction:
+            fullPath = os.path.join(self.root, path)
```

Comment:
I don't think that it matters, because there are no undo-able operations that happen before it.  Given that I'm going to refactor these to use `@transactional`, it will remain inside the transaction.